tools/serverpod_cli/lib/src/analyzer/entities/entity_parser/entity_parser.dart

```diff
@@ -0,0 +1,308 @@
+import 'package:serverpod_cli/src/util/extensions.dart';
+import 'package:serverpod_cli/src/util/protocol_helper.dart';
+import 'package:serverpod_cli/src/util/yaml_docs.dart';
+import 'package:serverpod_cli/src/generator/types.dart';
+import 'package:source_span/source_span.dart';
+import 'package:yaml/yaml.dart';
+
+import '../converter/converter.dart';
+import '../definitions.dart';
+import '../validation/keywords.dart';
+import '../validation/validate_node.dart';
+
+class EntityParser {
+  static SerializableEntityDefinition? serializeClassFile(
+    String documentType,
+    ProtocolSource protocolSource,
+    String outFileName,
+    YamlMap documentContents,
+    YamlDocumentationExtractor docsExtractor,
+    ValidateNode fieldStructure,
+  ) {
+    YamlNode? classNode = documentContents.nodes[documentType];
+
+    if (classNode == null) {
+      throw ArgumentError(
+        'No $documentType node found, only valid to call this function if '
+        ' the documentType exists as a top level key in the document.',
+      );
+    }
+
+    var classDocumentation = docsExtractor.getDocumentation(
+      documentContents.key(documentType)!.span.start,
+    );
+
+    var className = classNode.value;
+    if (className is! String) return null;
+
+    var tableName = _parseTableName(documentContents);
+    var serverOnly = _parseServerOnly(documentContents);
+    var fields = _parseClassFields(
+      documentContents,
+      docsExtractor,
+      tableName != null,
+      fieldStructure,
+    );
+    var indexes = _parseIndexes(documentContents, fields);
+
+    return ClassDefinition(
+      yamlProtocol: protocolSource.yaml,
+      className: className,
+      sourceFileName: protocolSource.yamlSourceUri.path,
+      tableName: tableName,
+      fileName: outFileName,
+      fields: fields,
+      indexes: indexes,
+      subDirParts: protocolSource.protocolRootPathParts,
+      documentation: classDocumentation,
+      isException: documentType == Keyword.exceptionType,
+      serverOnly: serverOnly,
+    );
+  }
+
+  static SerializableEntityDefinition? serializeEnumFile(
+    String documentType,
```

Comment:
Suggestion: I feel like this should probably not be a String type, given it's naming. Perhaps an enum with a method to convert the type to a string further down would be a better choice?

Reply:
I see your point, but this would imply we have to rethink the entire Keyword structure we have right now. They are all required to be used as string at some point though. 

Current Keyword class looks like this:

```dart
class Keyword {
  static const String classType = 'class';
  static const String exceptionType = 'exception';
  static const String enumType = 'enum';

  static const String serverOnly = 'serverOnly';
  static const String table = 'table';
  static const String fields = 'fields';
  static const String indexes = 'indexes';
  static const String values = 'values';

  static const String type = 'type';
  static const String unique = 'unique';
  static const String parent = 'parent';
  static const String api = 'api';
  static const String database = 'database';

  /// Special keyword to allow keys to be any string.
  static const String any = 'any';
}
```

Reply:
I'm thinking maybe a better name for the variable would work. E.g. `documentTypeName` or something with "keyword" in it?

Reply:
Hmm you know what for that specific method we shouldn't even have it there as it is not actually used!

Reply:
Fair! Pushed the change.